include/unifex/create.hpp

```diff
@@ -0,0 +1,196 @@
+namespace _create {
+
+template <typename Receiver, typename Fn, typename Context>
+struct _op {
+  struct type {
+    explicit type(Receiver rec, Fn fn, Context ctx)
+        noexcept(std::is_nothrow_move_constructible_v<Receiver> &&
+          std::is_nothrow_move_constructible_v<Fn> &&
+          std::is_nothrow_move_constructible_v<Context>)
+      : rec_((Receiver&&) rec), fn_((Fn&&) fn), ctx_((Context&&) ctx) {}
+
+    template (typename... Ts)
+      (requires receiver_of<Receiver, Ts...>)
+    void set_value(Ts&&... ts) noexcept(is_nothrow_receiver_of_v<Receiver, Ts...>) {
+      unifex::set_value((Receiver&&) rec_, (Ts&&) ts...);
+    }
+
+    template (typename Error)
+      (requires receiver<Receiver, Error>)
+    void set_error(Error&& error) noexcept {
+      unifex::set_error((Receiver&&) rec_, (Error&&) error);
+    }
+
+    void set_done() noexcept {
+      unifex::set_done((Receiver&&) rec_);
+    }
+
+    template (class Ctx = Context)
+      (requires (!same_as<Ctx, detail::_empty<>>))
+    Context const& context() const & noexcept { return ctx_; }
+
+    template (class Ctx = Context)
+      (requires (!same_as<Ctx, detail::_empty<>>))
+    Context&& context() && noexcept { return (Context&&) ctx_; }
+
+  private:
+    friend void tag_invoke(tag_t<start>, type& self) noexcept try {
+      self.fn_(self);
+    } catch(...) {
+      unifex::set_error((Receiver&&) self.rec_, std::current_exception());
+    }
+
+    // Forward other receiver queries
+    template (typename CPO)
+      (requires is_receiver_query_cpo_v<CPO> AND is_callable_v<CPO, const Receiver&>)
+    friend auto tag_invoke(CPO cpo, const type& self)
+        noexcept(is_nothrow_callable_v<CPO, const Receiver&>)
+      -> callable_result_t<CPO, const Receiver&> {
+      return std::move(cpo)(self.rec_);
+    }
+
+    UNIFEX_NO_UNIQUE_ADDRESS Receiver rec_;
+    UNIFEX_NO_UNIQUE_ADDRESS Fn fn_;
+    UNIFEX_NO_UNIQUE_ADDRESS Context ctx_;
+  };
+};
+
+template <typename Receiver, typename Fn, typename Context>
+using _operation = typename _op<Receiver, Fn, Context>::type;
+
+template <typename Fn, typename Context>
+struct _snd_base {
+  struct type {
+    template <template<typename...> class Variant>
+    using error_types = Variant<std::exception_ptr>;
+
+    static constexpr bool sends_done = false;
+
+    template (typename Self, typename Receiver)
+      (requires derived_from<remove_cvref_t<Self>, type> AND
+        constructible_from<Fn, member_t<Self, Fn>>)
+    friend _operation<remove_cvref_t<Receiver>, Fn, Context>
+    tag_invoke(tag_t<connect>, Self&& self, Receiver&& rec)
+        noexcept(std::is_nothrow_constructible_v<
+          _operation<Receiver, Fn, Context>,
+          member_t<Self, Fn>,
+          member_t<Self, Context>,
+          Receiver>) {
+      return _operation<remove_cvref_t<Receiver>, Fn, Context>{
+        (Receiver&&) rec,
+        (member_t<Self, Fn>&&) self.fn_,
+        (member_t<Self, Context>&&) self.ctx_};
+    }
```

Comment:
Can you explain what the `member_t` alias is doing here?  I've been staring at the definition and trying to cross-reference it to these uses and I can't figure it out.

Reply:
If `c` is an object of class type such that `decltype(c)` is `C`, _and_ the class has a data member `m` of type `M`, then `member_t<C, M>` is `decltype(c.m)`. If `Self` is not a reference type above, then it would be wrong to pass `self.fn_` to the `_operation` constructor because that would be an lvalue reference, inhibiting the move. So we cast to `member_t<Self, Fn>&&` which is an rvalue if `Self` is an rvalue reference or a non-reference type, and and lvalue if `Self` is an lvalue reference type.

Moves members when possible, copies when you need them.

Reply:
Nice.

Is that different from
```
static_cast<Self&&>(self).fn_
```
in this case?

Reply:
Ah, no it's not. That would probably be simpler.